Fragment being SubComponent of Activity example.

jemshit opened the pull request:
2 Fragment classes, its Component classes, layouts

jemshit commented:
To new branch if possible

jemshit committed:
Fragment being SubComponent of Activity example.
2 Fragment classes, its Component classes, layouts

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/fragment/FragmentScope.java:
```diff
@@ -0,0 +1,31 @@
+@Scope
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD, ElementType.TYPE})
+public @interface FragmentScope {
```
Given the usage of this scope we don't even need this. The fragment component can stay unscoped. I would recommend we only try to scope things that are expensive to provide or need to be ensured to only exist in a single instance for the component.

jemshit replied:
No need because of using SubComponent i guess 👍

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/fragment/MainActivityFragmentBinder.java:
```diff
@@ -0,0 +1,44 @@
+        FragmentOneComponent.class,
+        FragmentTwoComponent.class
+})
+public abstract class MainActivityFragmentBinder {
```
Split this into two classes `FragmentOneBinder.java` & `FragmentTwoBinder.java`. Then the activity component will include those binders rather than creating a binder specifically for it.

jemshit replied:
Actually your suggestion is another Use Case. I assumed those two Fragments will not be used again anywhere, thats why they are under same file with Specific Name: "MainActivityFragmentBinder". If we assume Fragment will be used in another Activity those two should be separated

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/fragment/fragment_one/FragmentOneComponent.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Module
+    class FragmentOneModule{
```
Lets make this its own class called something like `FragmentTextModule`. From there we can have both `FragmentOneComponent` & `FragmentTwoComponent` use the same module.

jemshit replied:
Yep, for this reusable example it is better to separate.

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/fragment/fragment_one/FragmentOneComponent.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Module
+    class FragmentOneModule{
+        String text;
```
Make this field `final`, it doesn't need to vary so let's prohibit it.

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/fragment/fragment_one/FragmentOneComponent.java:
```diff
@@ -0,0 +1,51 @@
+            this.text=text;
+        }
+
+        @Provides @FragmentScope @Named("TextFragmentOne")
```
Given that annotation are so cheap to create and String constants such as this spread throughout a code base can easily be messed up causing subtle bugs, let's go ahead and create a qualifier annotation such as `@FragmentText` to use instead of `@Named`.

They are easier to use too because autocomplete starts doing more work for you while consuming.

jemshit replied:
👍

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/fragment/fragment_two/FragmentTwo.java:
```diff
@@ -0,0 +1,67 @@
+
+    FragmentTwoComponent component;
+    @Inject @Named("TextFragmentTwo") String textFragmentTwo;
+    @Inject @Named("TextMainActivity") String textMainActivity;
```
Same deal on this `@Named(...)` let's remove it and create an `@ActivityText` qualifier annotation to use.

jemshit replied:
👍

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/fragment/fragment_two/FragmentTwo.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Override public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        component = ((FragmentComponentBuilderHost) getActivity())
```
I typically recommend that fragment injection be performed inside of `onCreateView` rather than `onCreate`. This is because `onCreate` is not always called depending on the path through the fragment lifecycle.

jemshit replied:
👍

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/fragment/fragment_two/FragmentTwoComponent.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Module
+    class FragmentTwoModule{
```
remove subclass and reuse as mentioned above.

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/main_activity/MainActivity.java:
```diff
@@ -16,22 +16,57 @@
+        textView= (TextView) findViewById(R.id.textView_mainActivity);
 
-        component = ((ActivityComponentBuilderHost) getApplication()).getActivityComponentBuilder(MainActivity.class, MainActivityComponent.Builder.class).build();
+        component = ((ActivityComponentBuilderHost) getApplication())
```
This is my bad since It was wrong before but will you please move the component creation and injection to happen before calling `setContentView(int)`?

This is because fragments can be included via xml and not always via in code fragment transactions. If we set content view first than any fragment included in the xml will cause a NPE when calling  `getFragmentComponentBuilder`

jemshit replied:
👍  indeed

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/main_activity/MainActivity.java:
```diff
@@ -16,22 +16,57 @@
+        getFragmentManager()
+                .beginTransaction()
+                .add(R.id.fragmentLayout_fragmentOne,new FragmentOne())
+                .add(R.id.fragmentLayout_fragmentTwo,new FragmentTwo())
```
would be nice to add fragment two via xml rather than here to illustrate that this strategy works either way.

jemshit replied:
Yeah, i was forced to do this because of "android.app.Fragment" in "FragmentComponentBuilder", now i moved fragments to xml and also updated inject-android lib version to 0.2.0

trevjonez commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,17 +1,43 @@
 <?xml version="1.0" encoding="utf-8"?>
 <RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
We can change this to a vertical linear layout

trevjonez commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,17 +1,43 @@
+        android:textSize="16sp"/>
+
+
+    <LinearLayout
```
this can be a frame layout

jemshit replied:
It is \<fragment\> now

trevjonez commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,17 +1,43 @@
+        tools:layout="@layout/fragment_one"/>
+
+    <LinearLayout
+        android:id="@+id/fragmentLayout_fragmentTwo"
```
as mentioned above it would be nice to include fragmentTwo via fragment tag

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/main_activity/MainActivityComponent.java:
```diff
@@ -16,18 +16,41 @@
+    }
+
+    @Module
+    class MainActivityModule{
```
Let's move this class to it's own file. I like to think of the component in terms of "Hey Dagger2 I need you to create this thing for me.", and the modules in terms of "I made this thing to help you do that stuff I asked for." Two separate and not necessarily related things.

And the biggest reason to do this is that some modules can be reused across different components for various reasons so separating it helps do that a bit more cleanly.

jemshit replied:
For reusability in this example i separated it. Normally i don't have reusable Modules of Activities, and i prefer Activity dependency codes in one place like this. Same can be done for View and Presenter interfaces like this example: https://www.novoda.com/blog/better-class-naming/

trevjonez commented on app/src/main/java/com/trevjonez/daggertest/main_activity/MainActivityComponent.java:
```diff
@@ -16,18 +16,41 @@
+
+    @Module
+    class MainActivityModule{
+        String text;
```
like before: `final`

trevjonez reviewed (changes requested):
Thanks for putting this all together. I'm fine with going straight into master with this if we can get some fairly simple changes made.

jemshit commented:
opening new pull request

jemshit committed:
@FragmentScope is removed because Fragment is already SubComponent of Activity and will be removed if Activity dies

jemshit committed:
@FragmentText, @ActivityText Qualifier annotations
ActivityTextModule, FragmentTextModule are new two reusable Module classes

jemshit committed:
Fragment dependency injection codes moved from onCreate() to onCreateView()

jemshit committed:
Activity dependency injection codes moved before setContentView()

jemshit committed:
Fragments are now from support.v4 lib
Fragments are added to Activity through xml only
Inject-Android library version is updated to 0.2.0

jemshit committed:
activity xml root layout is LinearLayout now
